src/devices/cddasongloader.cpp

```diff
@@ -47,18 +69,86 @@ QUrl CddaSongLoader::GetUrlFromTrack(int track_number) const {
 }
 
 void CddaSongLoader::LoadSongs() {
-  QtConcurrent::run(this, &CddaSongLoader::LoadSongsFromCdda);
+  if (!disc_.has_been_read || HasChanged()) {
+    if (!loading_future_.isRunning()) {
+      disc_.has_been_read = true;
+      loading_future_ =
+          QtConcurrent::run(this, &CddaSongLoader::LoadSongsFromCdda);
+    }
+  } else
+    emit SongsUpdated(disc_.tracks, false);
```

Comment:
`CddaDevice::Refresh` is called when the view asks for the items display text, though I'm not sure why:
https://github.com/clementine-player/Clementine/blob/master/src/devices/devicemanager.cpp#L194

That means that this signal could be emitted frequently.

Reply:
Hm. I'm not sure that's a problem, and if it is, I'd argue it's not one of `CddaSongLoader`. After all, if you ask me to load some songs for you, you shouldn't be surprised when you get a list of songs... I'll see if I can think of a way to make the `CddaDevice::Refresh` smarter (or maybe something even more upstream of that..)

Reply:
Yeah, the refresh looks like it was a bit of a hack.

This loader is different from the other device loaders (mtp, gpod, etc) in that its lifecycle is the same as the device object. The others are created, preform a single load, and then are deleted. And this is the pattern that SongLoader currently follows when using CddaSongLoader. So, maybe CddaSongLoader should be pared down and all of the synchronization should be handled by the device class instead? I think the cdio handle would need to be passed into CddaSongLoader in that case.

Reply:
That's a good point. I was going in the other direction so far (just pushed my pending changes): I have now extended CddaDevice to basically suppress SongsUpdated signals unless there was a substantial change. How this works now is that `CddaSongLoader::LoadSongs` will emit a `DiscChanged` signal if it detects a disc change before it starts reading the disc information. `CddaDevice` will then change an internal flag to accumulate all incoming `SongsUpdated` signals until it encounters one with `futher_updates_possible==false`, upon which it will emit its `SongsLoaded` signal and turn off the internal flag.
For calls to `Refresh` that do not result in a `DiscChanged` signal being received, `CddaDevice` ignores/suppresses the `SongsUpdated` signal it receives from `CddaSongLoader`.

It works now but I'm not really happy with it, because it increases the complexity of the signalling protocol of `CddaSongLoader`. Ideally I'd want that protocol to be stateless, i.e., `SongsUpdated` signals just happen and are self-contained - but that would require some very smart solution to the refreshing issue..

I'm actually thinking of absorbing the media change detection of PR #7024 into `CddaSongLoader` (or `CddaDevice`) and get rid off the `ConnectedDevice::Refresh` method entirely (its only used in `CddaDevice` anyways..). 

Another issue I've now run into is that the `cdio_get_media_changed` seems to not be working reliably when more than one `cdio` object is opened, which is now the case for the cd ripping (once held by `Ripper` and once by `CddaSongLoader`, both of which are used by `RipCDDialog`), which really messes with my changes now (which is another reason that motivates me to centralise all cdio related things in one class).

Thoughts?

Reply:
I wonder if the cdio instance references a singleton? One solution to that could be a static change counter that increments every time a change is detected and each instance could keep a local copy of the count last time it checked. Or maybe the CddaDevice instance can serve the ripper. I'm not sure how it works right now if you try to rip and play the CD at the same time.

I agree with getting rid of `ConnectedDevice::Refresh`. A timer in CddaDevice or CddaSongLoader makes more sense to me.

Also, doing this in #7024, or even a different PR, sounds like a good idea to me.

Reply:
> I wonder if the cdio instance references a singleton?
Looking at the code of the linux driver implementation of cdio, this doesn't seem to be the case. Each call to `cdio_open` creates a new instance and opens a new handle to the device. `cdio_get_media_changed` is just a wrapper around the `CDROM_MEDIA_CHANGED` ioctl call. Therefore, it seems that this might actually be a Linux kernel issue (and might not be an issue on other systems). I have somewhat confirmed this with a small test program that uses the ioctl call directly and could confirm that the call only returns 1 once when invoked for several separate file descriptors - this even holds across several independent processes.
I'll keep your suggestions in mind and try to reduce the number of cdio instances held by Clementine to one.